module/state_synchronization/execution_data_cid_cache_test.go

```diff
@@ -0,0 +1,120 @@
+package state_synchronization_test
+
+import (
+	"testing"
+
+	"github.com/ipfs/go-cid"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/state_synchronization"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+func TestCacheHit(t *testing.T) {
+	cache := state_synchronization.NewExecutionDataCIDCache(10)
+
+	header := unittest.BlockHeaderFixture()
+	var blobTree state_synchronization.BlobTree
+	blobTree = append(blobTree, []cid.Cid{unittest.CidFixture(), unittest.CidFixture()})
+	blobTree = append(blobTree, []cid.Cid{unittest.CidFixture()})
+	cache.Insert(&header, blobTree)
+
+	for height, cids := range blobTree {
+		for index, cid := range cids {
+			record, err := cache.Get(cid)
+			assert.NoError(t, err)
+			assert.Equal(t, record.BlobTreeRecord.BlockHeight, header.Height)
+			assert.Equal(t, record.BlobTreeRecord.BlockID, header.ID())
+			assert.Equal(t, record.BlobTreeLocation.Height, uint(height))
+			assert.Equal(t, record.BlobTreeLocation.Index, uint(index))
+		}
+	}
+}
+
+func TestCacheMiss(t *testing.T) {
+	cache := state_synchronization.NewExecutionDataCIDCache(10)
+
+	_, err := cache.Get(unittest.CidFixture())
+	assert.Error(t, err)
+}
+
+func TestCacheEviction(t *testing.T) {
+	size := uint(5)
+	cache := state_synchronization.NewExecutionDataCIDCache(size)
+
+	var headers []*flow.Header
+	var cids []cid.Cid
+
+	for i := uint(0); i < 2*size; i++ {
+		header := unittest.BlockHeaderFixture()
+		headers = append(headers, &header)
+		root := unittest.CidFixture()
+		cids = append(cids, root)
+		cache.Insert(&header, [][]cid.Cid{{root}})
+
+		for j, c := range cids {
+			record, err := cache.Get(c)
+			h := headers[j]
+
+			if len(cids)-j <= int(size) {
+				// cid should be in cache
+				assert.NoError(t, err)
+				assert.Equal(t, record.BlobTreeRecord.BlockHeight, h.Height)
+				assert.Equal(t, record.BlobTreeRecord.BlockID, h.ID())
+				assert.Equal(t, record.BlobTreeLocation.Height, uint(0))
+				assert.Equal(t, record.BlobTreeLocation.Index, uint(0))
+			} else {
+				// cid should be evicted
+				assert.ErrorIs(t, err, state_synchronization.ErrCacheMiss)
+			}
+		}
+	}
+}
+
+func TestDuplicateCID(t *testing.T) {
+	cache := state_synchronization.NewExecutionDataCIDCache(5)
+
+	c := unittest.CidFixture()
+	header1 := unittest.BlockHeaderFixture()
+	header2 := unittest.BlockHeaderFixture()
+
+	cache.Insert(&header1, [][]cid.Cid{{c}})
+	cache.Insert(&header2, [][]cid.Cid{{c}})
+
+	record, err := cache.Get(c)
+	assert.NoError(t, err)
+	assert.Equal(t, record.BlobTreeRecord.BlockHeight, header2.Height)
+	assert.Equal(t, record.BlobTreeRecord.BlockID, header2.ID())
+	assert.Equal(t, record.BlobTreeLocation.Height, uint(0))
+	assert.Equal(t, record.BlobTreeLocation.Index, uint(0))
```

Comment:
i guess theres no way to check size of the cache?

Reply:
I can add a method for that